### DISCUSSION TITLE
Need to parse the query with fragments

### DISCUSSION THREAD

**Discussion by gobimcp** [graphql-java/graphql-java#2604] [Q&A]
I am parsing the query to find out various fields and query variables in the query and I  came across fragments. Is there an API to get the final query after merging all the fragments?

**Comment by bbakerman** [comment]
> I am parsing the query to find out various fields 

yes there is - we are actively building it out right now - so it exists but its not fully public API

let me explain why 

Getting the pure Query AST nodes is a really difficult thing to get right - we are graphql experts and we have gotten it not quite right multiple times.

So we have gone back to the drawing board and started to rebuilt out a purpose built representation of the executable query and its types.

This led to `graphql.normalized.ExecutableNormalizedOperation` and its contains `graphql.normalized.ExecutableNormalizedField` - this reprsents a fully parsed query that has resolved all the fragments and has the fields and the types all resolved AND has the multiple choice "non concrete" types like interfaces and unions quasi resolved as well.

If you look this is not public API - yet - but we aim to get it there some release (but not the next)

Now all that said - what is that you want to do such that you need to know about all fields in a query.

We might be able to point you into the right direction via public API

**Reply by gobimcp** [reply]
@bbakerman Thanks for your response. Our use case is that we will have a query and we need to parsed metadata to build a UI layer for our product's users to input the values for query variables and create new filters, modified the selected list of properties etc.

For example if the query is 
```
query MyQuery($unique_value: String!) {
  MYSQL {
    myObject1Collection(filter: {
      my_unique_property_eq: $unique_value
    }) {
      items {
        id
        name
        description
        price
      }
    }
  }
  ORACLE {
    myObject2Collection(filter: {
      my_unique_property_eq: $unique_value
    }) {
      items {
        id
        name
        description
        price
      }
    }
  }
}
```

then we need the parsed metadata something like 
```
{
  "metadata": [
    {
      "dataSource": {
        "dataSourceType": "MYSQL",
        "dataSourceId": "myObject1Collection",
      },
      "filters": [
        {
          "name": "my_unique_property_eq",
          "type": "String",
          "isVariableValue": true,
          "variableName": "unique_value"
        }
      ],
      "properties": [
        "id",
        "name"
        "description"
        "price"
      ]
    },
    {
      "dataSource": {
        "dataSourceType": "ORACLE",
        "dataSourceId": "myObject2Collection",
      },
      "filters": [
        {
          "name": "my_unique_property_eq",
          "type": "String",
          "isVariableValue": true,
          "variableName": "unique_value"
        }
      ],
      "properties": [
        "id",
        "name"
        "description"
        "price"
      ]
    }
  ],
  "queryVariables": {
    "name": "unique_value",
    "type": "String",
    "isMandatory": true
  }
}
```

I figured out this public API https://github.com/graphql-java/graphql-java/blob/master/src/main/java/graphql/analysis/QueryTraverser.java in the code and I thought I can use this since it has schema type information for each field while traversing the query. But few things/issues I noticed are:

- The documentation says `It also doesn't consider field merging, which means for example { user{firstName} user{firstName}} will result in four visitField calls.` but when we have a fragment and if we use that fragment in more than one place in the query, this traverser actually visits the children of the fragment definition only once. But we need to traverse the children of fragment every time we encounter that fragmentSpread. (I somehow managed to achieve what I wanted by maintaining a separate deque to store the data I needed while parsing fragemntSpread and FragmentDefinition)
- This traverser requires variables values to be set while instantiating a new traverser. But I will just have the query at the time when I need to parse this, we do not have variables in hand. So I have hit a wall here using this QueryTraverser.java now..

Any suggestions will be really helpful.
